eip-5564.md

```diff
@@ -23,197 +22,103 @@ The standardization of non-interactive stealth address generation holds the pote
 ```
 
-### `IERC5564Messenger`
+The implementation of these methods is scheme-specific. The specification of a new stealth address scheme MUST specify the implementation for each of these methods. Additionally, although these function interfaces are specified in Solidity, they do not necessarily ever need to be implemented in Solidity, but any library or SDK conforming to this specification MUST implement these methods with compatible function interfaces.
+
+A one byte integer is used to identify stealth address schemes. A mapping from the scheme ID to it's specification is maintained in **{TODO where to maintain it?}**. An EIP is required to standardize a new stealth address scheme. These EIP extensions MUST specify:
+
+- The integer identifier for the scheme.
+- The algorithm for encoding a stealth meta-address (i.e. the spending public key and viewing public key) into a `bytes` array, and decoding it from `bytes` to the native key types of that scheme.
+- The algorithm for the `generateStealthAddress` method.
+- The algorithm for the `checkStealthAddress` method.
+- The algorithm for the `computeStealthKey` method.
+
+This specification additionally defines a singleton `ERC5564Messenger` contract that emits events to announce when something is sent to a stealth address. This MUST be a singleton contract, with one instance per chain. The contract is specified as follows:
 
 ```solidity
-/// @notice Interface for announcing that something was sent to a stealth address.
-interface IERC5564Messenger {
+/// @notice Interface for announcing when something is sent to a stealth address.
+contract IERC5564Messenger {
   /// @dev Emitted when sending something to a stealth address.
   /// @dev See the `announce` method for documentation on the parameters.
-  event Announcement(
-    bytes ephemeralPubKey, address indexed stealthAddress, bytes32 viewTag, bytes32 metadata
-  );
+  event Announcement(address indexed stealthAddress, bytes ephemeralPubKey, bytes32 metadata);
 
   /// @dev Called by integrators to emit an `Announcement` event.
-  /// @param ephemeralPubKey Ephemeral public key used by the sender.
   /// @param stealthAddress The computed stealth address for the recipient.
-  /// @param viewTag The hash of the shared secret used to compute the stealth address.
+  /// @param ephemeralPubKey Ephemeral public key used by the sender.
   /// @param metadata An arbitrary field that the sender can use however they like, but the below
   /// guidelines are recommended:
-  ///   - When sending ERC-20 tokens, the metadata SHOULD include the token address as the first 20
-  ///     bytes, and the amount being sent as the following 32 bytes.
+  ///   - The first byte of the metadata MUST be the stealth address scheme ID.
+  ///   - The second byte of the metadata MUST be the view tag.
+  ///   - When sending ERC-20 tokens, the metadata SHOULD include the token address as the next 20
+  ///     bytes, and the amount being sent in the remaining 10 bytes.
```

Comment:
At this point we could say that we use 1 byte for the methodID and 9 bytes for the amount. Then, users can distinguish between approvals and transfers (and almost every function since the probability that two functions start with the same byte are 1/256) and 9 bytes for the amount are sufficient.

Reply:
Another idea is that for tokens, reduce the precision used in the amount, so you can fit larger amounts by cutting off a few decimal places

Reply:
you're right. let's keep the amount with 10 bytes.
I'm not a fan of maintaining another mapping for that, especially since we can not foresee the upcoming function names in future contract and it would then always require some discussion if a certain function is worth it to be included in such a mapping or not.
Let's just keep it outside and let it be the job of the parser user to figure out which interation happend. the prasing provider may be able to dicrely tell the user if it was a transfer/approval etc. The options are quite small.

Reply:
yeah, I guess this will be an implementation detail. We could mention both, use the 10 bytes with 18 decimal precision also with less deciimals .f.e 1e-9.